### PR TITLE
Extended tracer

### PR DESCRIPTION
**Pull request by zeitlinger**
- Copy extended tracer from https://github.com/open-telemetry/opentelemetry-java-contrib/tree/main/extended-tracer
- use ExtendedTracer instead of Tracing and implement Tracer interface

Resolves https://github.com/open-telemetry/opentelemetry-java/issues/4610

### CODE REVIEW

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java:6]: There's a `io.opentelemetry.extension.propagation` package which would probably be more appropriate.

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java:63]: Let's return an immutable copy of this.

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java:50]: What does this comment do?

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java:46]:
```suggestion
   * metadata of an message.
```

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java:72]: Is there a reason this isn't public?

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java:68]:
```suggestion
   * message you're processing.
```

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java:85]: I'm not sure its appropriate to do this normalization here. @open-telemetry/java-instrumentation-approvers do we normalize before extracting context in `opentelemetry-java-instrumentation`?

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/README.md
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/README.md:1]:
```suggestion
# ExtendedTracer
```

File: extensions/incubator/src/test/java/io/opentelemetry/extension/incubator/trace/ExtendedTracerTest.java
Comment (jack-berg) [inline extensions/incubator/src/test/java/io/opentelemetry/extension/incubator/trace/ExtendedTracerTest.java:32]:
```suggestion
class ExtendedTracerTest {
```

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java:186]: I don't think this syntactic sugar is worth the API surface area.

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java:20]: This is the preferred pattern for creating instances in this repo, versus creating constructors with different overloads. The public static methods allow us to assign names that assist in conveying the intent / difference when multiple methods for creating exist.

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java:227]: I'm not particularly fond of the constructors for this class..
- The name ExtendedTracer suggests we might only need a tracer, but we actually need an OpenTelemetry as well
- The `ExtendedTracer(OpenTelemetry openTelemetry, String instrumentationScopeName)` constructor encourages creating a tracer without a version or schemaUrl. 

If these methods were to accept `OpenTelemetry` as an argument (or maybe even `ContextPropagators` or `TextMapPropagators`), then the construction can simplify to being a pure extension of `Tracer`, and only accept a `Tracer` delegate as an argument.

Presumably the goal of methods on ExtendedTracer in the incubator package is to someday have them added to `Tracer`. `Tracer` will never have access to an `OpenTelemetry` instance unless one is provided as an argument, so I think its better to reflect that in the API design of ExtendedTracer as well.

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/README.md
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/README.md:153]: Looks like these examples need to be updated to reflect the current naming. Let's hold off for now until we nail down the API, but remember to update before we merge.

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/SpanCallback.java
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/SpanCallback.java:16]:
```suggestion
public interface SpanCallable<T, E extends Throwable> {
```
Why the reference to spring docs in the javadoc?

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java
Comment (zeitlinger) [reply extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java:186]: are you suggesting to remove this feature altogether - or just that the user should call the static method in `Propagation` instead?

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java
Comment (trask) [reply extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java:227]: > or maybe even `ContextPropagators` or `TextMapPropagators`

👍

**Comment by jack-berg** [comment]
This would resolve #4610, correct @zeitlinger?

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java
Comment (zeitlinger) [reply extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java]: - provide propagators as params instead of otel instance
- use factory method instead of ctor

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java
Comment (zeitlinger) [reply extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java:75]: I think now that it's in a separate class I can make it public.

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java
Comment (zeitlinger) [reply extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java:6]: moved to `io.opentelemetry.extension.incubator.propagation`

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java
Comment (zeitlinger) [reply extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java:53]: IntelliJ complained about something... but not any more :smile:

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java
Comment (zeitlinger) [reply extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/Propagation.java]: good idea!

**Comment by zeitlinger** [comment]
@jack-berg I hope I've incorporated every feedback as intended.

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/README.md
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/README.md:1]: Let's move this to the root of `/extensions/incubator`

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/README.md
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/README.md:32]: This
```suggestion
ExtendedTracer extendedTracer = ExtendedTracer.create(tracer);
```

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java:211]: A couple of thoughts I don't feel overly strong about:
- What about a rename to `extractAndCall`? The difference between this functionality and the `call` methods is that context is extracted first, and the name of the method doesn't really imply that.
- What about switching `ContextPropagators` to the first argument? The argument order would be `propagators, carrier, spanBuilder, spanCallable, handleException`, which matches the argument order of the `call` family of methods and would allow for consistent overloads.
- What do you think about having a single method instead of overloads for the different span kinds? Not sure the overloads are that much easier than providing a span builder with the span kind set, and increase the surface area quite a bit.

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/SpanCallable.java
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/SpanCallable.java:16]:
```suggestion
  T callInSpan() throws E;
```

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/SpanRunnable.java
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/SpanRunnable.java:15]:
```suggestion
  void runInSpan() throws E;
```

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/propagation/Propagation.java
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/propagation/Propagation.java:25]: I'm tempted to call this `ExtendedContextPropagators` or `ContextPropagatorUtil`. WDYT?

**Review by jack-berg — COMMENTED** [review]
Just a couple of additional comments. I'm aligned at this point.

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/propagation/Propagation.java
Comment (zeitlinger) [reply extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/propagation/Propagation.java:25]: good idea

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java
Comment (zeitlinger) [reply extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java:211]: I really like it. Here's what I suggest:

```java
 public static <T, E extends Throwable> T extractAndCall(
      ContextPropagators propagators,
      Map<String, String> carrier,
      SpanBuilder spanBuilder,
      SpanKind spanKind,
      SpanCallable<T, E> spanCallable)
      throws E {
    return extractAndCall(propagators, carrier, spanBuilder, spanKind, spanCallable,  ExtendedTracer::setSpanError);
  }
```

- guides the user to remember span kind
- there's still an overload `handleException`, but that's only an extra arg

If we want to make it even better, we can use a builder. I think I'll try that out.

Comment (zeitlinger) [reply extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java:211]: Here's how it would roughly look like - I actually like it:

```java
                        t.extract(
                            otelTesting.getOpenTelemetry().getPropagators(),
                            Collections.emptyMap()).spanKind(SpanKind.CONSUMER).exceptionHandler(ignoreException).named("span").call(c)
```

Comment (zeitlinger) [reply extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java]: I've incorporated all other changes, but here I'm waiting for some feedback before implementing the builder pattern.

Comment (zeitlinger) [reply extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java]: @jack-berg wdyt?

Comment (zeitlinger) [reply extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java]: The example above is now

`                        t.spanBuilder("span")
                            .setSpanKind(SpanKind.SERVER)
                            .extractAndCall(
                                otelTesting.getOpenTelemetry().getPropagators(),
                                Collections.emptyMap(),
                                c)`

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/README.md
Comment (zeitlinger) [reply extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/README.md]: done

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/README.md
Comment (zeitlinger) [reply extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/README.md]: updated now

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedSpanBuilder.java
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedSpanBuilder.java:169]: I kind of like this approach, but don't like the `setExceptionHandler` method. Generally, I think it would be good to avoid having `ExtendedSpanBuilder` have any state besides the delegate SpanBuilder. 

I recommend:
- Removing setException handler and having it as an argument on the call / run methods.
- Rename `call` / `run` to `startAndCall` / `startAndRun`. This makes it clear that these methods are extended versions of the `start` method. The fact that `call` / `run` starts / ends the span is kind of surprising if you don't look at the javadoc / code.
- Rename `extractContext` to `extractAndSetParent`. Change the implementation to call `setParent` with the extracted context.

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedSpanBuilder.java
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedSpanBuilder.java:26]: #nit
```suggestion
public final class ExtendedSpanBuilder implements SpanBuilder {
```

File: extensions/incubator/src/test/java/io/opentelemetry/extension/incubator/trace/ExtendedTracerTest.java
Comment (jack-berg) [inline extensions/incubator/src/test/java/io/opentelemetry/extension/incubator/trace/ExtendedTracerTest.java:85]: #nit consider static import for `OpenTelemetryAssertions.equalTo` / `OpenTelemetryAssertions.satisfies` to improve readiability.

File: extensions/incubator/src/test/java/io/opentelemetry/extension/incubator/trace/ExtendedTracerTest.java
Comment (jack-berg) [inline extensions/incubator/src/test/java/io/opentelemetry/extension/incubator/trace/ExtendedTracerTest.java:139]: I believe this is redundant. `hasTracesSatisfyingExactly` should assert that size is equal to the number of consumer arguments.

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java:60]: These methods seem like unnecessary semantic sugar now.

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java
Comment (jack-berg) [inline extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java:91]: This method seems out of place. Not sure what to do about it.

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java
Comment (zeitlinger) [reply extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java:60]: I was also undecided - I'll remove it.

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedSpanBuilder.java
Comment (zeitlinger) [reply extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedSpanBuilder.java:169]: Are you suggesting not to end the span in `startAndCall`? This would kind of defeat the purpose.

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java
Comment (zeitlinger) [reply extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java:91]: as discussed, let's scope this out into a separate PR

**Review by jack-berg — APPROVED** [review]

File: extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java
Comment (zeitlinger) [reply extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java]: Here it is: https://github.com/open-telemetry/opentelemetry-java/pull/6063

**Pull request merged**
